### PR TITLE
refactor(core): allow passing passive option to `addEventListener()`

### PR DESCRIPTION
**Pull request by jeantimex**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/main/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [x] Documentation content changes
- [ ] angular.dev application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?

Issue Number: N/A

## What is the new behavior?
This PR makes event library's `addEventListener()` calls `EventTarget.addEventListener()` with `options` instead of `useCapture`. This allows us to pass or override `passive` option for improving scroll performance in future PRs.

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

## Other information
 - Passing `options` to `EventTarget.addEventListener` is not supported in IE11
 - `removeEventListener()` doesn't recognize `passive` option

### CODE REVIEW

File: packages/core/primitives/event-dispatch/src/event.ts [hunk 1]
```diff
@@ -64,7 +64,17 @@ export function addEventListener(
...
   }
-  element.addEventListener(eventType, handler, capture);
+
+  const options: AddEventListenerOptions = {capture};
+
+  // For passive events, set passive option to `false` to:
+  // 1. Allow preventDefault() to be called.
+  // 2. Prevent browser violation warnings about passive event listeners.
+  if (isPassiveEventType(eventType)) {
+    options.passive = false;
```
Comment (rahatarmanahmed) [inline packages/core/primitives/event-dispatch/src/event.ts:74]: While well intentioned, I would avoid explicitly setting `passive = false`. This will get in the way of browser optimizations that default passive to true when unspecified.

File: packages/core/primitives/event-dispatch/src/event.ts [hunk 2]
```diff
@@ -64,7 +64,17 @@ export function addEventListener(
   if (isCaptureEventType(eventType)) {
     capture = true;
   }
-  element.addEventListener(eventType, handler, capture);
+
+  const options: AddEventListenerOptions = {capture};
+
+  // For passive events, set passive option to `false` to:
+  // 1. Allow preventDefault() to be called.
+  // 2. Prevent browser violation warnings about passive event listeners.
```
Comment (rahatarmanahmed) [inline packages/core/primitives/event-dispatch/src/event.ts:72]: Just to be sure there's not a misunderstanding here: Browsers should be warning when you have a **non-passive** event listener for the 4 event types in this CL, right? This seems backwards.

**Review by rahatarmanahmed — COMMENTED** [review]
As mentioned in the comments, we don't need to explicitly set passive to false. Can we change this PR to just pass through the passive option and only set it if it's set to true? I think it's fine to do in one PR.

(optionally we could only use the options object form of the capture/passive argument when passive is needed so that it doesn't break older browsers that don't support it, but we don't support such older browsers anyway.)

**Comment by jeantimex** [comment]
I can add an addtional parameter to the addEventListener function, if capture or passive is set in the options, I will use them in EventTarget.addEventListener().

File: packages/core/primitives/event-dispatch/src/event.ts (hunk 1 above)
Comment (jeantimex) [reply packages/core/primitives/event-dispatch/src/event.ts]: I updated `addEventListener()` to allow passing an optional event options.

File: packages/core/primitives/event-dispatch/src/event.ts (hunk 2 above)
Comment (jeantimex) [reply packages/core/primitives/event-dispatch/src/event.ts]: For touchstart, touchmove, wheel and mousewheel, when we don't set passive explicitly, browsers will give the violation warning.

I update this PR to allow us to pass event options to `addEventListener()`, and makes sure it doesn't introduce breaking changes.

File: packages/core/primitives/event-dispatch/src/event.ts [hunk 3]
```diff
@@ -38,12 +38,14 @@ export function getBrowserEventType(eventType: string) {
...
  * @param eventType The event type.
  * @param handler The handler function to install.
+ * @param eventOptions An object that specifies characteristics about the event listener.
  * @return Information needed to uninstall the event handler eventually.
  */
 export function addEventListener(
   element: Element,
   eventType: string,
   handler: (event: Event) => void,
+  options?: AddEventListenerOptions,
```
Comment (rahatarmanahmed) [inline packages/core/primitives/event-dispatch/src/event.ts:48]: I don't think we want to let jsaction users to enable the capture listeners. Some of the [other options](https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/addEventListener#options) also seem possibly hard to support. Can we limit this to just the `passive` option?

Comment (jeantimex) [reply packages/core/primitives/event-dispatch/src/event.ts:48]: Sure, let me make the change to only support `passive` option.

Comment (jeantimex) [reply packages/core/primitives/event-dispatch/src/event.ts]: Done.

File: packages/core/primitives/event-dispatch/src/event.ts [hunk 4]
```diff
@@ -79,7 +86,15 @@ export function addEventListener(
  */
 export function removeEventListener(element: Element, info: EventHandlerInfo) {
   if (element.removeEventListener) {
-    element.removeEventListener(info.eventType, info.handler as EventListener, info.capture);
+    if (typeof info.passive === 'boolean') {
```
Comment (rahatarmanahmed) [inline packages/core/primitives/event-dispatch/src/event.ts:89]: According to [this](https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/removeEventListener#matching_event_listeners_for_removal), only the capture option matters for removeEventListener. So we can revert this change and also don't need to track the passive option in EventHandlerInfo either.

**Review by rahatarmanahmed — COMMENTED** [review]
It's not super useful to just edit this one method. You'll want to pipe the passive argument through these two functions, too:

- https://github.com/angular/angular/blob/085e447cf8a1c7dc7b1e9c04a6531bf72d1446b2/packages/core/primitives/event-dispatch/src/eventcontract.ts#L177
- https://github.com/angular/angular/blob/085e447cf8a1c7dc7b1e9c04a6531bf72d1446b2/packages/core/primitives/event-dispatch/src/event_contract_container.ts#L69

**Comment by rahatarmanahmed** [comment]
> It's not super useful to just edit this one method. You'll want to pipe the passive argument through these two functions, too:
> 
> * https://github.com/angular/angular/blob/085e447cf8a1c7dc7b1e9c04a6531bf72d1446b2/packages/core/primitives/event-dispatch/src/eventcontract.ts#L177
> * https://github.com/angular/angular/blob/085e447cf8a1c7dc7b1e9c04a6531bf72d1446b2/packages/core/primitives/event-dispatch/src/event_contract_container.ts#L69

Oh, once you've done that could you also update the documentation here about the new option?: https://github.com/angular/angular/blob/085e447cf8a1c7dc7b1e9c04a6531bf72d1446b2/packages/core/primitives/event-dispatch/README.md

File: packages/core/primitives/event-dispatch/src/event.ts (hunk 4 above)
Comment (jeantimex) [reply packages/core/primitives/event-dispatch/src/event.ts:89]: Sounds good!

**Comment by jeantimex** [comment]
> > It's not super useful to just edit this one method. You'll want to pipe the passive argument through these two functions, too:
> > 
> > * https://github.com/angular/angular/blob/085e447cf8a1c7dc7b1e9c04a6531bf72d1446b2/packages/core/primitives/event-dispatch/src/eventcontract.ts#L177
> > * https://github.com/angular/angular/blob/085e447cf8a1c7dc7b1e9c04a6531bf72d1446b2/packages/core/primitives/event-dispatch/src/event_contract_container.ts#L69
> 
> Oh, once you've done that could you also update the documentation here about the new option?: https://github.com/angular/angular/blob/085e447cf8a1c7dc7b1e9c04a6531bf72d1446b2/packages/core/primitives/event-dispatch/README.md

Done. I pass `passive` through event contract's `addListener()` all the way to event lib's `addEventListener()`. Let me know if that's what you expect, if it looks good to you, I will then update the documentation. Thank you!

**Comment by rahatarmanahmed** [comment]
Looks good. You'll need to patch this into google3 and run a TGP to make sure it doesn't break anything. I'll ping you instructions

**Comment by jeantimex** [comment]
Updated the documentation to mention the new `passive` argument in `addEvent` function.

**Review by rahatarmanahmed — APPROVED** [review]

**Comment by jeantimex** [comment]
TAP Train passed, here is the [result](https://fusion2.corp.google.com/presubmit/686203800/OCL:686203800:BASE:686566399:1729101333044:adfe2166;groups=Pending/targets).

File: packages/core/primitives/event-dispatch/src/event.ts [hunk 5]
```diff
@@ -64,9 +67,14 @@ export function addEventListener(
...
     capture = true;
   }
-  element.addEventListener(eventType, handler, capture);
 
-  return {eventType, handler, capture};
+  if (typeof passive === 'boolean') {
+    element.addEventListener(eventType, handler, {capture, passive});
+  } else {
+    element.addEventListener(eventType, handler, capture);
+  }
```
Comment (AndrewKushnir) [inline packages/core/primitives/event-dispatch/src/event.ts:75]: We can probably save some bytes (and similarly for the removeEventListener fn below):
```suggestion
  const options = typeof passive === 'boolean' ? {capture, passive} : capture;
  element.addEventListener(eventType, handler, options);
```

**Review by AndrewKushnir — COMMENTED** [review]
LGTM, just a small proposal to minimize code size.

File: packages/core/primitives/event-dispatch/src/event.ts (hunk 5 above)
Comment (jeantimex) [reply packages/core/primitives/event-dispatch/src/event.ts]: Done.

File: packages/core/primitives/event-dispatch/src/event.ts [hunk 6]
```diff
@@ -79,7 +84,13 @@ export function addEventListener(
...
 export function removeEventListener(element: Element, info: EventHandlerInfo) {
   if (element.removeEventListener) {
-    element.removeEventListener(info.eventType, info.handler as EventListener, info.capture);
+    if (typeof info.passive === 'boolean') {
+      element.removeEventListener(info.eventType, info.handler as EventListener, {
+        capture: info.capture,
+      });
+    } else {
+      element.removeEventListener(info.eventType, info.handler as EventListener, info.capture);
+    }
```
Comment (AndrewKushnir) [inline packages/core/primitives/event-dispatch/src/event.ts:93]: If I understand correctly, this change is note required - both `removeEventListener` functions in each branch would do exactly the same.

Comment (jeantimex) [reply packages/core/primitives/event-dispatch/src/event.ts:93]: According to [Matching event listeners for removal](https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/removeEventListener#matching_event_listeners_for_removal), "It's worth noting that some browser releases have been inconsistent on this, and unless you have specific reasons otherwise, it's probably wise to use the same values used for the call to addEventListener() when calling removeEventListener()."

Comment (AndrewKushnir) [reply packages/core/primitives/event-dispatch/src/event.ts:93]: Thanks for additional information. I believe with both options we deliver the same input into the removal logic, so I don't see a reason to use an object in one case and a boolean flag in another. We should check/confirm though that we can still remove listeners via `removeEventListener(..., true)`, which were originally added via `addEventListener(..., {capture: true})`.

@rahatarmanahmed would be curious to hear your opinion too.

Comment (rahatarmanahmed) [reply packages/core/primitives/event-dispatch/src/event.ts:93]: Hard to say w/o knowing which browser versions were "inconsistent" and in which way they were inconsistent. I don't think there's a good way to find out so either option is fine to me.

Comment (AndrewKushnir) [inline packages/core/primitives/event-dispatch/src/event.ts:93]: I'd propose a similar refactoring here + also add a comment on why we pass capture like this (we can use a quote from https://github.com/angular/angular/pull/58198/files#r1809081367):
```suggestion
    const options = typeof info.passive === 'boolean' ? {capture: info.capture} : info.capture;
    element.removeEventListener(info.eventType, info.handler as EventListener, options);
```

Comment (AndrewKushnir) [reply packages/core/primitives/event-dispatch/src/event.ts:93]: Sounds good, let's keep it safe, I've added a quick comment here as well: https://github.com/angular/angular/pull/58198#pullrequestreview-2385491157.

Comment (jeantimex) [reply packages/core/primitives/event-dispatch/src/event.ts]: Sounds good! I updated my PR and added a comment :)

Comment (jeantimex) [reply packages/core/primitives/event-dispatch/src/event.ts]: Done.

**Review by rahatarmanahmed — APPROVED** [review]

**Review by AndrewKushnir — APPROVED** [review]

**Comment by jeantimex** [comment]
I was trying to squash all the commits in my branch, but ended up with a lot of changes :(

**Comment by jeantimex** [comment]
Let me close this PR in favor of https://github.com/angular/angular/pull/58316
